File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/util/ConfigStructureAnalysis.java
package me.bristermitten.mittenlib.annotations.util;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;
import java.util.HashSet;
import java.util.Set;
import javax.lang.model.type.TypeMirror;
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure;
import me.bristermitten.mittenlib.annotations.compile.ConfigNameCache;

@Singleton
public class ConfigStructureAnalysis {
    private final ConfigNameCache cache;

    @Inject
    public ConfigStructureAnalysis(ConfigNameCache cache) {
        this.cache = cache;
    }

    /**
     * Returns whether this {@link AbstractConfigStructure} is dynamically initializable, i.e.
     * every property is either nullable or has a default value.
     */
    public boolean isDynamicallyInitializable(AbstractConfigStructure structure) {
        return isDynamicallyInitializable(structure, new HashSet<>());
    }

    /**
     * Returns whether this {@link TypeMirror} is a configuration type that MittenLib knows how to initialize.
     */
    public boolean isTypeInitializable(TypeMirror type) {
        return isTypeInitializable(type, new HashSet<>());
    }

    private boolean isDynamicallyInitializable(AbstractConfigStructure structure, Set<ClassName> visited) {
        if (!visited.add(structure.name())) {
            return true; // Cycle detected, assume true to allow other properties to decide
        }
        return structure.properties().stream()
                .allMatch(p -> p.settings().hasDefaultValue()
                        || p.settings().isNullable()
                        || isTypeInitializable(p.propertyType(), visited));
    }

    private boolean isTypeInitializable(TypeMirror type, Set<ClassName> visited) {
        return cache.lookupAST(type)
                .map(structure -> isDynamicallyInitializable(structure, visited))
                .orElse(false);
    }

    public boolean needsValidation(AbstractConfigStructure structure) {
        return structure.properties().stream()
                .anyMatch(p -> !p.settings().constraints().isEmpty()
                        || !p.settings().elementConstraints().isEmpty()
                        || !p.settings().keyConstraints().isEmpty()
                        || (!TypeName.get(p.propertyType()).isPrimitive()
                                && !p.settings().isNullable()));
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/util/ConfigStructureAnalysis.scala
package me.bristermitten.mittenlib.annotations.util

import com.google.inject.{Inject, Singleton}
import com.palantir.javapoet.ClassName
import me.bristermitten.mittenlib.annotations.ast.{
  AbstractConfigStructure,
  Property => AstProperty
}
import me.bristermitten.mittenlib.annotations.compile.ConfigNameCache
import me.bristermitten.mittenlib.annotations.domain.*

import javax.lang.model.`type`.DeclaredType
import javax.lang.model.`type`.TypeMirror
import javax.lang.model.element.TypeElement

@Singleton
class ConfigStructureAnalysis @Inject() (
    private val cache: ConfigNameCache,
    private val typesUtil: TypesUtil
):

  def isDynamicallyInitializable(structure: ConfigStructure): Boolean =
    isDynamicallyInitializable(structure, Set.empty)

  def isDynamicallyInitializable(structure: AbstractConfigStructure): Boolean =
    cache.lookupDomain(structure.name()).exists(isDynamicallyInitializable)

  private def isDynamicallyInitializable(
      structure: ConfigStructure,
      visited: Set[ClassName]
  ): Boolean =
    val name = structure.name
    if (visited.contains(name)) true
    else
      val nextVisited = visited + name
      structure.properties.forall { p =>
        p.hasDefault || p.isNullable || isTypeInitializable(
          p.propertyType,
          nextVisited
        )
      }

  def isTypeInitializable(tpe: PropertyType): Boolean =
    isTypeInitializable(tpe, Set.empty)

  def isTypeInitializable(tpe: TypeMirror): Boolean =
    if (typesUtil.isConfigType(tpe)) {
      tpe match {
        case declaredType: DeclaredType =>
          val className = ClassName.get(
            declaredType.asElement().asInstanceOf[TypeElement]
          )
          val opt = cache.lookupDomain(className)
          opt.exists(isDynamicallyInitializable(_, Set.empty))
        case _ => false
      }
    } else {
      typesUtil.isOptional(tpe)
    }

  def hasDefaultOrIsInitializable(p: Property): Boolean =
    p.hasDefault || isTypeInitializable(p.propertyType)

  def hasDefaultOrIsInitializable(p: AstProperty): Boolean =
    p.settings().hasDefaultValue() || isTypeInitializable(p.propertyType())

  private def isTypeInitializable(
      tpe: PropertyType,
      visited: Set[ClassName]
  ): Boolean = tpe match {
    case PropertyType.ConfigProperty(className, _) =>
      val opt = cache.lookupDomain(className)
      if (opt.isDefined) isDynamicallyInitializable(opt.get, visited)
      else false
    case PropertyType.OptionalProperty(_) => true
    case _                                => false
  }

  def needsValidation(structure: ConfigStructure): Boolean =
    structure.properties.exists { p =>
      hasConstraints(p.propertyType) || (!isPrimitive(
        p.propertyType
      ) && !p.isNullable)
    }

  def needsValidation(structure: AbstractConfigStructure): Boolean =
    cache.lookupDomain(structure.name()).exists(needsValidation)

  private def hasConstraints(pt: PropertyType): Boolean = pt match {
    case PropertyType.Primitive(_, cs)       => cs.nonEmpty
    case PropertyType.Object(_, cs)          => cs.nonEmpty
    case PropertyType.ListProperty(elem, cs) =>
      cs.nonEmpty || hasConstraints(elem)
    case PropertyType.Map(key, value, cs) =>
      cs.nonEmpty || hasConstraints(key) || hasConstraints(value)
    case PropertyType.Set(elem, cs) => cs.nonEmpty || hasConstraints(elem)
    case PropertyType.ConfigProperty(_, cs)  => cs.nonEmpty
    case PropertyType.EnumType(_, _, cs)     => cs.nonEmpty
    case PropertyType.OptionalProperty(elem) => hasConstraints(elem)
  }

  private def isPrimitive(pt: PropertyType): Boolean = pt match {
    case PropertyType.Primitive(_, _) => true
    case _                            => false
  }
